### ISSUE TITLE
Feature: Option to make test run fail if there is a 404

### ISSUE THREAD

**Issue by miyasudokoro** [karma-runner/karma#3442]
### Expected behaviour
If there is a 404, the Karma process should (optionally) fail.

### Actual behaviour
If there is a 404, it is just logged. 

### Environment Details

- Karma version (output of `karma --version`): all
- Relevant part of your `karma.config.js` file

### Steps to reproduce the behaviour
- Specify a file that does not exist in the karma.conf file.
- Import a file that does not exist into a source file or a spec file.
- Incorrectly configure your proxies.
- Put a typo in one of the files you are trying to list.
- Rename a specified file but fail to update the config file.

The main reason this is an issue is that it allows failures to be silent. In my case, I have correctly configured my proxies, so every time there is a 404, it means that there is a mistake in my source code. It is not good CI/CD to have to manually scroll through my test output to check for 404s rather than having the test simply fail. 

I think, ideally, this is an option that would take arrays of patterns, similar to "exclude". For example, a user might want to have only files matching the patterns "*.js" or "*.css" throw an error on 404, but allow image files to have a 404 without consequence. In my case, I have correctly configured my proxies, so I would want any and every 404 to result in an error -- if any of them occur, it means that there is a typo in my source code, and I want to know about it.

**Comment by johnjbarton** [comment]
Sounds great!

404 warning is here:
https://github.com/karma-runner/karma/blob/131d154dd7a873ae536da628b94b4418cfb5b35f/lib/middleware/common.js#L86
And the caller has the `config` object here the `failOn404` pattern could be accessed:
https://github.com/karma-runner/karma/blob/131d154dd7a873ae536da628b94b4418cfb5b35f/lib/middleware/common.js#L68

The new unit test would live here:
https://github.com/karma-runner/karma/blob/5f13e11784c283872c97501b9e21402987429c56/test/unit/web-server.spec.js#L202

**Comment by miyasudokoro** [comment]
My company's policy does not allow me to contribute to open source projects that are in use by my company, so I'm afraid I can't make a pull request myself. Many thanks to anyone who can pick this up.

**Comment by TimvdLippe** [comment]
We hit the same issue in DevTools. I will take a punt at this.

**Comment by TimvdLippe** [comment]
Sadly the middleware does not have access to the exit code handling. I think the Karma client needs to listen for 404's and report an error instead. Strangely enough, Karma should have errored on 404's already: https://github.com/karma-runner/karma/blob/3b854adfad717a22613eee1223f4c92f3f4b1e5a/context/karma.js#L75-L79

However, as it turns out, script elements that 404 don't actually cause the `window.onerror` to run. They run their own `onerror` and call it a day. Of course, catching only these 404's wouldn't catch proxy issues.

Therefore @johnjbarton could you let me know what the nicest way is for the middleware to communicate to the runner to produce an error? Essentially what I would like to do is something like `this.emit('404')` in the `middleware/common.js` and that the runner picks that up and correctly exists with code 1.

**Comment by johnjbarton** [comment]
Sorry I don't follow.  Is the 404 we are discussing "karma does not have a route for a url and thus must reply with 404"?  The of course the middleware can solve this problem.  Or is the 404 "user test code contacts random non-existent URL which karma knows nothing about and never sees because it is running in a different process"?  Maybe you can guess I'm skeptical about solving that one ;-)  Or maybe a third option?

**Comment by TimvdLippe** [comment]
Yes the former. Essentially what I want to do is error if the following happens:

1. I specify in Karma that I want to run test files A, B, and C.
2. While A and B are valid test files, C is not. E.g. C does not exist on the filesystem
3. The middleware detects that C does not exist and currently warns on the console. For example, these are our logs: https://logs.chromium.org/logs/devtools-frontend/buildbucket/cr-buildbucket.appspot.com/8881724298040167536/+/steps/Unit_Tests/0/stdout (The `does not match any file`) error.
4. While Karma reports it as a warning, the exit code is 0 (because all the tests passed) and we didn't block the CL from being submitted.

Looking at the error again, I now see that it is not the middleware server that reports the warning, but the `file-list` implementation. This is still relevant for the original request in OP (`Specify a file that does not exist in the karma.conf file.`), but is not the `middleware/common.js` I thought was reporting the warning.

Maybe we can allow Karma to exit with code 1 for this use case first and later deal with the `middleware` trying to serve an unknown file?

**Comment by TimvdLippe** [comment]
I turned it around and built the failure logic in our `karma.conf.js` as an interim solution: https://chromium-review.googlesource.com/c/devtools/devtools-frontend/+/2174424

It doesn't catch the 404's we would otherwise see with for example images, but it is a stopgap for the "this test is missing" issue. Hopefully it is useful to others who are looking for a solution.

**Comment by miyasudokoro** [comment]
I had not been back to this issue for a while, and I am a bit surprised it hasn't been fixed yet. @TimvdLippe I'm not sure what you are talking about when you say there's no access to the exit code handling. Can you clarify? Is there some child process that I don't know about that also needs killing?

This has worked for me for 6 months now:

````
// anywhere where there is access to the config
process.env.FAIL_ON_404 = !!config.failOn404;

// inside karma/lib/middleware/common.js
function serve404 (response, path) {
  if ( process.env.FAIL_ON_404 ) {
    log.error(`404: ${path}`)
    process.exitCode = 1
    return process.exit()
  }
  // the existing code goes here
}
````

I just pop this IF statement in every time I reinstall Karma. I actually set process.env.FAIL_ON_404 = true all the way up in my karma.conf.js file for my convenience rather than using the config at all. 

By the way, unlike what you described, my main reason for requesting this feature is actually because I sometimes introduce typos in my spec file imports:

````
import myModule from '/whoops/a/typo';  <-- reported as 404: /base/src/whoops/a/typo due to my proxy settings

// this line is never reached so no tests fail
describe( 'test', () => {
````

There is nothing I can do about this particular problem using karma.conf. The files are all present, after all.
